Updated and simplified info on Smart Rollups

timothymcmackin opened the pull request:
Simplified info on Smart Rollups to what makes sense for Tezos developers on docs.tezos.com. Preview: https://docs-staging-git-architecture-overview-smart-rollups-trili-tech.vercel.app/architecture/smart-rollups

The PR includes this new diagram: https://lucid.app/documents/embedded/3953825e-e5aa-43bb-bc43-701bb0ac3b73#

timothymcmackin commented on docs/architecture/smart-rollups.md:
```diff
@@ -1,1310 +1,147 @@
+The code of a Smart Rollup is public, and anyone can run a Smart Rollup node to run that code and verify that other nodes are running it correctly, just like anyone can run baking nodes and accuser nodes on layer 1.
 
-This page covers an advanced topic at the bleeding edge of Tezos core development. If you are interested in more fundamental reading, a great place to start is [Tezos Protocol and Shell](./protocol) and [Smart Contracts](../smart-contracts/languages).
+For an introduction to Smart Rollups, see the tutorial [Deploy a Smart Rollup](../tutorials/smart-rollup).
```
Clarified the tutorial link and added a link to the Octezz reference on Smart Rolups.

timothymcmackin commented on docs/architecture/smart-rollups.md:
```diff
@@ -1,1310 +1,147 @@
-the prefix `sr1`.
+Smart Rollups allow you to run large amounts of processing that would be impractical on layer 1.
+For example, Smart Rollups are not constrained to the layer 1 block baking rate and therefore can run far more transactions per second than layer 1.
+Smart Rollups can also manipulate large amounts of data without needing to store it on layer 1.
```
Added "They also allow you to avoid transaction fees and storage fees."

timothymcmackin commented on docs/architecture/smart-rollups.md:
```diff
@@ -1,1310 +1,147 @@
+Smart Rollups can also manipulate large amounts of data without needing to store it on layer 1.
 
-## Inputs
+Smart Rollups also enable [Etherlink](https://www.etherlink.com/), which connects Tezos to Ethereum.
```
Changed this to "For example, Smart Rollups enable [Etherlink](https://www.etherlink.com/), which makes it possible to run EVM applications (originally written for Ethereum) on Tezos."

timothymcmackin commented on docs/architecture/smart-rollups.md:
```diff
@@ -1,1310 +1,147 @@
+The reveal data channel supports these requests:
 
-This mode uses [Wasmer](https://wasmer.io) when running WebAssembly code which allows you to configure the compiler it will use to deal with the WebAssembly code. It can be done using the `OCTEZ_WASMER_COMPILER` environment variable which will be picked up by the smart rollup node.
+- A rollup node can request arbitrary data up to 4KB at a time if it knows the blake2b hash of the data, known as _preimage requests_.
```
"- A rollup node can request an arbitrary data page up to 4KB at a time if it knows the blake2b hash of the page, known as _preimage requests_."

timothymcmackin commented on docs/architecture/smart-rollups.md:
```diff
@@ -1,1310 +1,147 @@
 
-Once again, note that values from the memory are output as is, since the representation is internal to WASM.
+Each node must stake 10,000 tez along with its commitment.
+When another node makes an identical commitment, its stake is added to the stake of the first commitment.
```
"Nodes must stake 10,000 tez along with their commitments. When nodes make identical commitments, their stakes are combined into a single stake for the commitment."

timothymcmackin commented on docs/architecture/smart-rollups.md:
```diff
@@ -1,1310 +1,147 @@
-either automatically or through specific commands. The debugger can automatically fill preimages from files in a specific directory on the disk, by default in the `preimage` subdirectory of the working directory. It can be configured with the option `--preimage-dir <directory>`.
+This refutation game happens as many times as is necessary to eliminate incorrect commitments.
+Because the node that ran the PVM correctly is guaranteed to win the refutation game, a single honest node is enough to ensure that the Smart Rollup is running correctly.
+For this reason, Smart Rollups are sometimes called Smart Optimistic Rollups.
```
"This kind of Smart Rollup is called a Smart Optimistic Rollup because the commitments are assumed to be correct until they are proven wrong by an honest rollup node."

timothymcmackin commented on docs/architecture/smart-rollups.md:
```diff
@@ -1,1310 +1,147 @@
+For this reason, Smart Rollups are sometimes called Smart Optimistic Rollups.
 
-In case there is no corresponding file found for the requested preimage, the debugger will ask for the hexadecimal value of the preimage:
+At the end of the refutation period, the correct commitment is said to be _cemented_ and therefore final and unchangeable.
```
"and becomes final and unchangeable"

timothymcmackin commented on docs/architecture/smart-rollups.md:
```diff
@@ -1,1310 +1,147 @@
 ---
 
-Rollups play a crucial part in providing next-generation scaling on Tezos. This page gives a technical introduction to Smart Rollups, their optimistic nature, and an intro to developing your own WASM kernel.
+Smart Rollups handle transactions and logic in a separate environment that doesn't need to follow all of the same rules as layer 1.
```
I changed it around to not talk about layers before we define what layers are:

> Smart Rollups play a crucial part in providing high scalability on Tezos. They handle transactions and logic in a separate environment that doesn't need to follow all of the same rules as the main Tezos network. For this reason, they are not subject to the same block delays and data space constraints as the main Tezos network.
> 
> The transactions and logic that Smart Rollups run is called _layer 2_ to differentiate it from the main network, which is called _layer 1_.

NicNomadic reviewed (changes requested):
Besides the various remark, I have the general remark that there is a lot of repetition in section "Smart Rollup lifecycle" with respect to the previous parts. Of course, some amount of redundancy is not banned, but the final section should sound more like a recap than like repeating the steps with about the same amount of details.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,192 @@
+## Uses for Smart Rollups
+
+Smart Rollups allow you to run large amounts of processing that would be impractical on layer 1.
+They also allow you to avoid some transaction fees and storage fees.
```
> Smart Rollups allow you to run large amounts of processing that would be too slow or expensive to run on layer 1. Thus they allow you to avoid some transaction fees and storage fees.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,192 @@
+Smart Rollups can also manipulate large amounts of data without needing to store it on layer 1.
+
+Smart Rollups can also retrieve data from outside the blockchain in specific ways that smart contracts can't do.
+For example, Smart Rollups enable [Etherlink](https://www.etherlink.com/), which makes it possible to run EVM applications (originally written for Ethereum) on Tezos.
```
I was considering etherlink as a way to get information from EVM and put it on Tezos, but that is an oversimplification. How about this:

> Smart Rollups can also provide execution environments that are compatible with other blockchains. For example, Smart Rollups enable [Etherlink](https://www.etherlink.com/), which makes it possible to run EVM applications (originally written for Ethereum) on Tezos.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,192 @@
+  - revealing data from the (WIP) DAL
+*/}
+
+## Consensus
```
Reworked Consensus section into Commitments.

timothymcmackin replied:
Later I removed the Commitment section and merged it with the lifecycle section to reduce duplication.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,192 @@
+
+Smart Rollup nodes achieve consensus differently from layer 1 nodes.
+Instead of constantly sharing blocks and updating their states, layer 2 nodes maintain their states internally and share them at predefined intervals.
+Because the nodes behave in a deterministic manner, their states should all be the same if they have processed the same Smart Rollup inbox messages.
```
> Because Smart Rollup nodes behave in a deterministic manner, their states should all be the same if they have processed the same Smart Rollup inbox messages with the same kernel starting from the same origination level.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,192 @@
+- The _commitment period_ runs for 60 consecutive blocks.
+During the commitment period, the rollups receive messages from the Smart Rollup inbox, process them, and update their internal states.
+
+  At the end of the commitment period, Smart Rollup nodes use the PVM to generate a hash of their states and publish it to layer 1.
```
> At the end of the commitment period, each node generates a commitment and publishes it to layer 1.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,192 @@
+  Then a new commitment period begins.
+
+- The _refutation period_ for a commitment period starts when the commitment period ends.
+During the refutation period, if two or more commitments are different, the nodes play a _refutation game_ to identify the correct commitment.
```
> if the commitments are not all identical

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,192 @@
+- The _refutation period_ for a commitment period starts when the commitment period ends.
+During the refutation period, if two or more commitments are different, the nodes play a _refutation game_ to identify the correct commitment.
+
+  Then the correct commitment is said to be _cemented_ and therefore final and unchangeable.
```
> When there is only one commitment left, either because all nodes published identical commitments or the refutation game eliminated all incorrect commitments, this correct commitment is said to be _cemented_ and therefore final and unchangeable.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,192 @@
+1. During the commitment period, the Smart Rollup nodes receive the messages in the Smart Rollup inbox, filter them to the messages that they are interested in, and run processing based on those messages.
+1. The nodes add outgoing messages to the Smart Rollup outbox in the form of calls to smart contracts.
+1. At the end of the commitment period, the nodes publish their commitments to layer 1.
+These commitments are proofs generated from the state of the Smart Rollup node, which means that if all of the nodes run the Smart Rollup correctly, the commitments are all the same.
```
This sentence repeats info from the Commitment section so I removed it.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,192 @@
+1. At the end of the commitment period, the nodes publish their commitments to layer 1.
+These commitments are proofs generated from the state of the Smart Rollup node, which means that if all of the nodes run the Smart Rollup correctly, the commitments are all the same.
+1. A new commitment period begins.
+1. At the same time as the new commitment period begins, a refutation period for the previous commitment period begins, during which time other nodes can verify the commitment and publish a concurrent commitment to confirm or refute it.
```
> At the same time as the new commitment period begins, a refutation period for the previous commitment period begins, during which time other nodes can publish a concurrent commitment to refute a published commitment.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,192 @@
+The nodes automatically play the refutation game by stepping through their logic to identify the point at which they differ.
+At this point, layer 1 uses the PVM to identify the correct commitment.
+Then the protocol gives half of the incorrect commitment's stake to the correct commitment's stake, and burns the other half.
+Then the protocol eliminates the incorrect commitment because it has no stake.
```
> The protocol eliminates the incorrect commitment...

timothymcmackin commented:
> Besides the various remark, I have the general remark that there is a lot of repetition in section "Smart Rollup lifecycle" with respect to the previous parts. Of course, some amount of redundancy is not banned, but the final section should sound more like a recap than like repeating the steps with about the same amount of details.

Thanks for the review. I have reduced some duplication by removing the commitment/consensus section and trimming the lifecycle steps. I'll look for other places to trim this down.

NicNomadic commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,171 @@
+This kind of Smart Rollup is called a Smart Optimistic Rollup because the commitments are assumed to be correct until they are proven wrong by an honest rollup node.
+
+When there is only one commitment left, either because all nodes published identical commitments or the refutation game eliminated all incorrect commitments, this correct commitment is said to be _cemented_ and therefore final and unchangeable.
+A commitment can't be cemented until the end of the refutation period and until all previous commitments are cemented.
```
1. This phrase and the one above are true, but the same things can be said in a simpler way by combining them in one phrase (note that a refutation game can go past the end of the refutation period; it must just *start* before this time):
2. "final and unchangeable" are not consequences of the *definition* of "cemented" (so you cannot use "therefore"); you should either make them part of the definition or introduce it as synonyms
"When there is only one commitment left, either because all nodes published identical commitments and the refutation period expired, or because the refutation game eliminated all incorrect commitments, this correct commitment is said to be _cemented_ and becomes final and unchangeable.

timothymcmackin replied:
> When there is only one commitment left, either because all nodes published identical commitments or the refutation game eliminated all incorrect commitments, this correct commitment is said to be _cemented_ and becomes final and unchangeable. A commitment can't be cemented all previous commitments are cemented. The commitments for the next commitment period build on this cemented commitment.

NicNomadic replied:
In fact, I'm not sure the first phrase is true. Cementation is not automatic, it's a L1 operation that must be injected by someone. Secondly, there may be several refutation games. Thirdly, we have to mention the refutation period has ended in the first case. And let's also remove the invalid sentence "A commitment can't be cemented all previous commitments are cemented.". This gives:

> When there is only one commitment left, either because all nodes published identical commitments during the whole refutation period or because this commitment won the refutation game(s) and eliminated all other  commitments, then this correct commitment can be cemented by a dedicated L1 operation and becomes final and unchangeable. The commitments for the next commitment period build on this cemented commitment.

timothymcmackin replied:
Updated. I didn't know that there was a specific layer 1 operation required to cement a commitment.

NicNomadic reviewed (changes requested):
As there were major changes, I did a second complete pass.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,171 @@
+Smart Rollups allow you to run large amounts of processing that would be too slow or expensive to run on layer 1.
+Thus they allow you to avoid some transaction fees and storage fees.
+For example, Smart Rollups are not constrained to the layer 1 block baking rate and therefore can run far more transactions per second than layer 1.
+Smart Rollups can also manipulate large amounts of data without needing to store it on layer 1.
```
"manipulating date" and "getting data from outside the blockchain" are different use cases, so I don't think they belong together. I think "manipulating large amounts of data" fits better in the paragraph about "run large amounts of processing that would be too slow or expensive to run on layer 1" because you can manipulate large amounts of data on layer 1, it would just be cost prohibitive.

How about this:

> Smart Rollups allow you to run large amounts of processing and manipulate large amounts of data that would be too slow or expensive to run on layer 1. Thus they allow you to avoid some transaction fees and storage fees. Also, Smart Rollups are not constrained to the layer 1 block baking rate and therefore can run far more transactions per second than layer 1.

NicNomadic replied:
> "manipulating [large] data" and "getting data from outside the blockchain" are different use casesI

Maybe, although the former is a consequence of the latter. OK, let's group this like you do. But then let's make 4 use cases:
1. heavy computations and/or large data
2. high transaction throughput
3. data from outside
4. implementing other execution environments

Then, all what you have to do is add a blank line in this paragraph before "Also" (and maybe remove the "Also").

timothymcmackin replied:
Made this into a bulleted list and added the "avoid fees" use case:

- Smart Rollups allow you to run large amounts of processing and manipulate large amounts of data that would be too slow or expensive to run on layer 1.

- Smart Rollups are not constrained to the layer 1 block baking rate and therefore can run far more transactions per second than layer 1.

- Smart Rollups allow you to avoid some transaction fees and storage fees.

- Smart Rollups can retrieve data from outside the blockchain in specific ways that smart contracts can't do.

- Smart Rollups can implement different execution environments, such as execution environments that are compatible with other blockchains. For example, Smart Rollups enable [Etherlink](https://www.etherlink.com/), which makes it possible to run EVM applications (originally written for Ethereum) on Tezos.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,171 @@
-Finally, after the application of the operations of the Tezos block, the layer 1 pushes one final internal message `"End of level"`. Similarly to `"Start of level"`, this internal messages does not come with any payload.
+### Rollup inbox
+
+As part of its state, layer 1 has a _rollup inbox_ that contains messages from layer 1 to all rollups.
```
I changed this to "as part of its state" because we wanted to describe the rollup inbox not as part of any individual block but as part of the L1 state.

NicNomadic replied:
What I'm saying is precisely that this is not correct. The L1 state (aka the L1 "context") contains things such as contract balances; but it does not contain the L2 inbox, AFAIU.

timothymcmackin replied:
Asked for clarification in a smart rollups slack channel.

timothymcmackin replied:
Per the smart rollups channel, it's more technically accurate to say that each layer 1 block has a smart rollup inbox. I'll change this to "Each layer 1 block contains a _rollup inbox_..."

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,171 @@
+Similarly, smart contracts can add messages in a way similar to calling a smart contract entrypoint, by using the Michelson `TRANSFER_TOKENS` instruction.
+The messages that smart contracts add to the inbox are called _internal messages_.
+
+Each block also contains these internal messages, which are created by the protocol:
```
I don't think it's necessary to say "automatically" or "predefined;" it's clear that these are messages that are in each block.

NicNomadic replied:
OK, then maybe just:
```suggestion
Each block also contains the following internal messages, which are created by the protocol:
```

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,171 @@
+During the refutation period for a commitment period, if two or more nodes publish different commitments, two of them play a _refutation game_ to identify the correct commitment.
+The nodes automatically play the refutation game by stepping through their logic to identify the point at which they differ.
+At this point, layer 1 uses the PVM to identify the correct commitment.
+Then the protocol gives half of the incorrect commitment's stake to the correct commitment's stake, and burns the other half.
```
What's the "any" in "if any?" Does this mean that perhaps both commitments are incorrect so if neither of them win the refutation game, 100% of both of their stakes are burned because there is no correct commitment to give it to?

NicNomadic replied:
Yes. In the reference doc:

> There are only two possible outcomes: either one of the staker has provided a valid proof, then that staker wins the game, and is rewarded with half of the opponent’s deposit (the other half being burnt); or, both stakers have provided an invalid proof and they both lose their deposit.

timothymcmackin replied:
Updated with this list, which is a little wordy but clear:

Each refutation game has one of two results:

- Neither commitment is correct.
In this case, the protocol burns both commitments' stakes and eliminates both commitments.

- One commitment is correct and the other is not.
In this case, the protocol eliminates the incorrect commitment, burns half of the incorrect commitment's stake, and gives the other half to the correct commitment's stake.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,171 @@
+The nodes automatically play the refutation game by stepping through their logic to identify the point at which they differ.
+At this point, layer 1 uses the PVM to identify the correct commitment.
+Then the protocol gives half of the incorrect commitment's stake to the correct commitment's stake, and burns the other half.
+The protocol eliminates the incorrect commitment because it has no stake.
```
> The protocol eliminates the incorrect commitment or commitments because they have no stake.

NicNomadic replied:
The commitment is not incorrect "because" it has no stake. It's rather the other way around. But as this has already been explained above, let's shorten it:
> The protocol eliminates the incorrect commitment or commitments.

timothymcmackin replied:
Addressed in a different comment.

NicNomadic commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,171 @@
+Nodes must stake 10,000 tez along with their commitments.
+When nodes make identical commitments, their stakes are combined into a single stake for the commitment.
+
+Then a refutation period for the commitment period begins, and a new commitment period begins.
```
In fact, the next commitment period has already started by the time the commitment is published, and this has been explained above. So let's get rid of that part:
```suggestion
Then a refutation period for the commitment period begins.
```

timothymcmackin replied:
Reworked based on these comments. Reduced some duplication in  this section.

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,171 @@
+1. Origination: A user originates the Smart Rollup to layer 1 and one or more users start nodes based on that Smart Rollup.
+1. Commitment periods: The Smart Rollup nodes receive the messages in the Smart Rollup inbox, run processing based on those messages, and publish a hash of their state at the end of the period.
+1. Refutation periods: Nodes can publish a concurrent commitment to refute a published commitment.
+1. Triggering outbox messages: After the commitment is cemented, any client can trigger operations from the Smart Rollup.
```
Changed this to "When the commitment can no longer be refuted..."

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,171 @@
+
+The general flow of a Smart Rollup goes through these phases:
+
+1. Origination: A user originates the Smart Rollup to layer 1 and one or more users start nodes based on that Smart Rollup.
```
Updated to "Origination: A user called the _rollup operator_ originates the Smart Rollup to layer 1 and one or more users start nodes based on that Smart Rollup to independently verify its operation."

timothymcmackin commented:
> As there were major changes, I did a second complete pass.

Thank you. I think I've addressed all comments except for [this question to Yann](https://github.com/trilitech/tezos-developer-docs/pull/163#discussion_r1430338890) and a question about how to describe the rollup inbox that I have posted to some experts in slack.

NicNomadic commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,177 @@
+Because Smart Rollup nodes behave in a deterministic manner, their states should all be the same if they have processed the same inbox messages with the same kernel starting from the same origination level.
+This state is referred to as the "state of the rollup."
+
+At the end of a commitment period, the next commitment period starts and at the same time a refutation period starts for the commitment period that just ended.
```
After discussion with Sylvain, it appears clearly that the refutation does not start at the same time as the next commitment period, but rather when the first commitment for current commitment period is published:
```suggestion
At the end of a commitment period, the next commitment period starts.
```

NicNomadic commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,177 @@
+
+Because the PVM is deterministic and all of the inputs are the same for all nodes, any honest node that runs the same Smart Rollup produces the same commitment.
+As long as nodes publish matching commitments, they continue running normally.
+
```
Introduce the refutation period here:
```suggestion

When the first commitment for a past commitment period is published, a *refutation period* starts, during which any rollup node can publish its own commitment for the same commitment period, especially if it did not achieve the same state.
```

NicNomadic reviewed (changes requested):
I found a few other important aspects to fix, please.

NicNomadic commented on docs/architecture/accounts.md:
```diff
@@ -18,4 +18,4 @@ See [Smart contracts](../smart-contracts).
 - Smart Rollup accounts are another type of originated account.
```
As you said just above that smart contracts = originated accounts, you cannot say here that smart rollups are another type of originated accounts. Maybe rather "programmable accounts", or "accounts containing code and storage"?

NicNomadic replied:
I don't think that smart rollups are a kind of originated account. In fact, they are not at all accounts. So I think:
- you should revert the change to smart contracts (which are the same thing as "originated account")
- either rename this list as "addresses" instead of "accounts", or remove smart rollups from the list

timothymcmackin replied:
I'm removing changes to this file from this PR and I'm working on clarifying accounts and addresses in this pr: https://github.com/trilitech/tezos-developer-docs/pull/252

acerone85 commented on docs/architecture/accounts.md:
```diff
@@ -11,11 +11,11 @@ Tezos uses these types of accounts:
 Any wallet application or the Octez command-line tool can create implicit accounts.
 
-- Smart contract accounts (also known as _originated accounts_) store immutable code, mutable storage, tez (ꜩ), and tickets.
+- Smart contract accounts (a type of _originated account_) store immutable code, mutable storage, tez (ꜩ), and tickets.
```
this implies that there are other types of originated accounts, which I don't think is true?

timothymcmackin replied:
Are Smart Rollup accounts not originated accounts?

NicNomadic replied:
As told in a different thread, I don't think they are. If we want to reclassify them as accounts, we have to discuss this in a documentation venue (e.g. EffortDoc).

timothymcmackin replied:
I'm removing changes to this file from this PR and I'm working on clarifying accounts and addresses in this pr: https://github.com/trilitech/tezos-developer-docs/pull/252

acerone85 commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,178 @@
-title: Smart Optimistic Rollups
-authors: 'Nomadic Labs, TriliTech'
+title: Smart Rollups
+authors: 'Nomadic Labs, TriliTech, Tim McMackin'
```
I'm not sure we list individual contributors as authors in documentation.

timothymcmackin replied:
We're not consistent about this; some files list organizations, other list individual authors.

timothymcmackin replied:
I'll straighten this out in the future, but for now it's not shown anywhere on the site and we can ignore it.

acerone85 commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,178 @@
 
-## Examples
+Smart Rollups play a crucial part in providing high scalability on Tezos.
+They handle transactions and logic in a separate environment that doesn't need to follow all of the same rules as the main Tezos network.
```
I find this sentence a bit vague. The important bit for smart rollups is that the processing of transactions is performed in a separate environment. Another bit that should be stressed here is that each rollup has its own logic for processing transactions.

Also the part about not following the rules as the main Tezos network might be misleading. Rather they have dedicated rules that are enforced by the main Tezos network.

timothymcmackin replied:
Because this is an introduction section, I'd like to keep it vague and answer the question "What are SCORUs and why would I want to use them?"

I could combine this sentence and the next one into: "They handle logic in a separate environment that can run transactions at a much higher rate and use larger amounts of data than the main Tezos network."

acerone85 commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,178 @@
-## Examples
+Smart Rollups play a crucial part in providing high scalability on Tezos.
+They handle transactions and logic in a separate environment that doesn't need to follow all of the same rules as the main Tezos network.
+For this reason, they are not subject to the same block delays and data space constraints as the main Tezos network.
```
This is also a bit misleading. 
Rollups by themselves are still subject to the same block delays of the L1 network. In order to be processed, data that allows to retrieve L2 transactions (either directly or by using a Data availability solution, see below) must be posted to the rollup inbox on the L1. 

Data space constraints require a data availability solution which is separate from rollups. Smart rollups, however, integrate with two data availability solutions: the Data Availability Layer (DAL) and Data Availability Committees (DACs), the latter via a mechanism called the reveal data channel.

timothymcmackin replied:
How about "They handle logic in a separate environment that can run transactions at a much higher rate and can use larger amounts of data than the main Tezos network."

acerone85 commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,178 @@
-For examples of Smart Rollups, see this repository: https://gitlab.com/tezos/kernel-gallery.
+The transactions and logic that Smart Rollups run is called _layer 2_ to differentiate it from the main network, which is called _layer 1_.
+
+The compiled code of a Smart Rollup must be provided publicly, so that anyone can run a Smart Rollup node to execute that code and verify that other nodes are running it correctly, just like anyone can run nodes, bakers, and accusers on layer 1.
```
In practice, because of the space constraints on the L1, only a small kernel (the installer kernel) is used to originate a rollup. This code is public as it is stored on the L1. The first action of the installer kernel is that of installing a larger kernel, whose code is imported using the reveal data channel. The blockchain itself does not guarantee the availability of this data, and instead it is the responsibility of the originator of the rollup to make it available to rollup operators.

timothymcmackin replied:
I think the installer kernel and kernel are too detailed for this intro, so I changed this sentence to:
> Anyone can run a node based on a Smart Rollup to execute its code and verify that other nodes are running it correctly, just like anyone can run nodes, bakers, and accusers on layer 1.

acerone85 commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,178 @@
+The transactions and logic that Smart Rollups run is called _layer 2_ to differentiate it from the main network, which is called _layer 1_.
+
+The compiled code of a Smart Rollup must be provided publicly, so that anyone can run a Smart Rollup node to execute that code and verify that other nodes are running it correctly, just like anyone can run nodes, bakers, and accusers on layer 1.
+This code, called the _kernel_, must behave in a deterministic manner and according to a given semantics, which guarantees that results are reproducible by any rollup node with the same kernel.
```
Deterministic execution is ensured by the reference PVM, so it is virtually impossible to write a kernel that behaves non-deterministically

timothymcmackin replied:
Changed to:
> This code, called the _kernel_, runs in a deterministic manner...

jobjo commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,178 @@
+
+- Smart Rollups allow you to run large amounts of processing and manipulate large amounts of data that would be too slow or expensive to run on layer 1.
+
+- Smart Rollups are not constrained to the layer 1 block baking rate and therefore can run far more transactions per second than layer 1.
```
This is not entirely true. Without a "sequencer",  a rollup receives input by following L1 blocks and hence are subject to the same baking rate.

timothymcmackin replied:
But they can run a lot of logic in between blocks, so they aren't constrained by putting every transaction into a block. Is there a way to explain how/why L2 nodes can run "far more TPS" or should I simplify this to say "Smart Rollups can run far more transactions per second..."?

timothymcmackin replied:
Changed to "Smart Rollups can run far more transactions per second than layer 1."

NicNomadic commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,177 @@
+The general flow of a Smart Rollup goes through these phases:
+
+1. Origination: A user called the _rollup operator_ originates the Smart Rollup to layer 1 and one or more users start nodes based on that Smart Rollup to independently verify its operation.
+1. Commitment periods: The Smart Rollup nodes receive the messages in the Smart Rollup inbox, run processing based on those messages, generate but do not run outbox messages, and publish a hash of their state at the end of the period.
```
As the next point talks about published commitments, you have to introduce it here:
```suggestion
1. Commitment periods: The Smart Rollup nodes receive the messages in the Smart Rollup inbox, run processing based on those messages, generate but do not run outbox messages, and publish a hash of their state at the end of the period., called a commitment
```

NicNomadic commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,177 @@
+Smart Rollups can generate transactions to run on layer 1, but those transactions do not run immediately.
+When a commitment includes layer 1 transactions, these transactions go into the Smart Rollup outbox and wait for the commitment to be cemented.
+
+After the commitment is cemented, clients can trigger transactions in the outbox with the Octez `execute outbox message` command.
```
```suggestion
After the commitment is cemented, clients can trigger transactions in the outbox with the Octez client `execute outbox message` command.
```

timothymcmackin commented on docs/architecture/smart-rollups.mdx:
```diff
@@ -1,266 +1,177 @@
+The semantics is precisely defined by a reference virtual machine called a proof-generating virtual machine (PVM), able to generate a proof that executing a program in a given context results in a given state.
+This guarantees that if a divergence in results is found, it can be tracked down to a single elementary step that was not executed correctly by some node.
+In this way, multiple nodes can run the same rollup and each node can verify the state of the rollup.
+
```
This sounds like an implementation detail for people running SRs that is more appropriate in the Octez docs.

NicNomadic replied:
I agree with the version you proposed on slack.

timothymcmackin replied:
Updated to read:

> Anyone can run a node based on a Smart Rollup to execute its code and verify that other nodes are running it correctly, just like anyone can run nodes, bakers, and accusers on layer 1. This code, called the kernel, runs in a deterministic manner and according to a given semantics, which guarantees that results are reproducible by any rollup node with the same kernel. The semantics is precisely defined by a reference virtual machine called a proof-generating virtual machine (PVM), able to generate a proof that executing a program in a given context results in a given state. During normal execution, the Smart Rollup can use any virtual machine that is compatible with the PVM semantics, which allows the Smart Rollup to be more efficient.
> 
> Using the PVM and optionally a compatible VM guarantees that if a divergence in results is found, it can be tracked down to a single elementary step that was not executed correctly by some node. In this way, multiple nodes can run the same rollup and each node can verify the state of the rollup.

NicNomadic replied:
thx

NicNomadic reviewed (commented):
LGTM
